File: array_hashing/EncodeDecodeString.java
//https://neetcode.io/problems/string-encode-and-decode

class Solution {
    /**
     * We will encode with length of the string + # + string
     */
    public String encode(List<String> strs) {
        StringBuilder res = new StringBuilder();

        for (String s: strs){
            res.append(s.length()).append('#').append(s);
        }
        return res.toString();

    }

    /**
     * First get the length. The length will be the integer value upuntill #
     * After the # to the length is a string
     */
    public List<String> decode(String str) {
        List<String> res = new ArrayList<>();

        int i = 0;
        while (i < str.length()){
            int j = i;
            while (str.charAt(j) != '#'){
                j++;
            }
            int length = Integer.parseInt(str.substring(i, j));
            i = j + 1;
            j = i + length;
            res.add(str.substring(i, j));
            i = j;
        }
        return res;
    }
}

File: array_hashing/encode_decode_string.py
# https://neetcode.io/problems/string-encode-and-decode


class Solution:

    def encode(self, strs: List[str]) -> str:
        enc_str = ""

        for s in strs:
            enc_str += str(len(s)) + "$" + s
        return enc_str

    def decode(self, s: str) -> List[str]:
        dec_strs = []

        i = 0

        while i < len(s):
            j = i

            while s[j] != "$":
                j += 1

            length = int(s[i:j])
            dec_strs.append(s[j + 1 : j + 1 + length])
            i = j + 1 + length
        return dec_strs
